octopus_sensing/devices/skinos_streaming.py

```diff
@@ -0,0 +1,264 @@
+            ddata = self._serial.read(1)
+
+    def _run(self):
+        '''
+        Listening to the message queue and manage messages
+        '''
+        loop_thread = threading.Thread(target=self._stream_loop)
+        loop_thread.start()
+
+        while True:
+            message = self.message_queue.get()
+            if message is None:
+                continue
+            if message.type == MessageType.START:
+                if self._state == "START":
+                    print("Skinos streaming has already recorded the START triger")
+                else:
+                    print("Skinos start")
+                    self._experiment_id = message.experiment_id
+                    self.__set_trigger(message)
+                    self._state = "START"
+            elif message.type == MessageType.STOP:
+                if self._state == "STOP":
+                    print("Skinos streaming has already recorded the STOP triger")
+                else:
+                    if self._saving_mode == SavingModeEnum.SEPARATED_SAVING_MODE:
+                        self._experiment_id = message.experiment_id
+                        file_name = \
+                            "{0}/{1}-{2}-{3}.csv".format(self.output_path,
+                                                        self.name,
+                                                        self._experiment_id,
+                                                        message.stimulus_id)
+                        self._save_to_file(file_name)
+                    else:
+                        print("Skinos stop")
+                        self._experiment_id = message.experiment_id
+                        self.__set_trigger(message)
+                    self._state = "STOP"
+            elif message.type == MessageType.TERMINATE:
+                if self._saving_mode == SavingModeEnum.CONTINIOUS_SAVING_MODE:
+                    file_name = \
+                        "{0}/{1}-{2}.csv".format(self.output_path,
+                                                 self.name,
+                                                 self._experiment_id)
+                    self._save_to_file(file_name)
+                break
+
+        self._break_loop = True
+        loop_thread.join()
+
+    def __set_trigger(self, message: Message):
+        '''
+        Takes a message and set the trigger using its data
+
+        Parameters
+        -----------
+        message: Message
+                 a message object
+        '''
+        self._trigger = \
+            "{0}-{1}-{2}".format(message.type,
+                                 message.experiment_id,
+                                 str(message.stimulus_id).zfill(2))
+
+    def _stream_loop(self):
+        '''
+        Reads incoming data
+        '''
+        initial_flag = False
+        framesize = 15
+
+        try:
+            while True:
+                time.sleep(0.1)
+
+                result = self._serial.read_until(b':', 16)
```

Comment:
The `read_until` is a blocker method. If for some reason the device doesn't return anything, we will never reach the `if self._break_loop` line, so a `TERMINATE` message won't be able to break the loop and stop the process.

It is very unlikely that this happen. But if you can fix it, it will make the code more solid and reliable.

I recommend adding a `time_out` here, and then checking if the result ends with the expected byte. If it wasn't, it means the timeout has happened and there's something wrong with the device. Then you break the loop, save the data, and terminate the process. Because the device is not working.

The timeout can be a long time, like five seconds, so if it happens, it means the device is not responsive.